### PR TITLE
let's read LoadCredentials=/SetCredentials= style cred in sysusers/firstboot and when asking for passwords

### PR DESCRIPTION
**Pull request by poettering**
Now we have this nice credentials concept, let's make us of it across the board.

1. let's use it in sysusers for configuring passwords for users that are created
2. let's use it for firstboot for the same usecase, plus for querying a few very basic other creds
3. let's use it in the "ask-password" logic to supply passwords whenever possible

In particular 1+2 are very useful in `systemd-nspawn --volatile=yes -b …` invocations since in that case /etc comes up empty and we cannot provision a root password beforehand hence. By adding `--set-credential=passwd.plaintext-password.root:foobar` we can tell firstboot/sysusers to set the root pw to "foobar" and use that for logging in afterwards.

### CODE REVIEW

**Comment by poettering** [comment]
(this is material for after v248, no need to hurry this in)

File: src/basic/creds-util.c [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+bool credential_name_valid(const char *s) {
+        /* We want that credential names are both valid in filenames (since that's our primary way to pass
+         * them around) and as fdnames (which is how we might want to pass them around eventually) */
+        return filename_is_valid(s) && fdname_is_valid(s);
+}
+
+int get_credentials_dir(const char **ret) {
+        const char *e;
+
```
Comment (bluca) [inline src/basic/creds-util.c:16]: assert(ret)

File: man/systemd.exec.xml [hunk 2]
```diff
@@ -2831,19 +2831,21 @@ StandardInputData=SWNrIHNpdHplIGRhIHVuJyBlc3NlIEtsb3BzLAp1ZmYgZWVtYWwga2xvcHAncy
...
+        to it (only once at unit start-up) and the credential data read from the connection, providing an
+        easy IPC integration point for dynamically providing credentials from other services. If the
+        specified path is not absolute and itself qualifies as valid credential identifier it is understood
+        to refer to a credential that the service manager itself received via the
+        <varname>$CREDENTIALS_DIRECTORY</varname> environment variable, which may be used to propagate
+        credentials from an invoking environment (e.g. a container manager that invoked the service manager)
+        into a service. The contents of the file/socket may be arbitrary binary or textual data, including
+        newline characters and <constant>NUL</constant> bytes. If the file system path is omitted it is
+        chosen identical to the credential name, i.e. this is a terse way do declare credentials to inherit
+        from the service mnager into a service. This option may be used multiple times, each time defining an
```
Comment (bluca) [inline man/systemd.exec.xml]: /mnager/manager/

**Comment by bluca** [comment]
The integration with nspawn is great - how about hosts? Is there any way/plan to allow preseeded firstboot? Maybe via the kernel command line?

**Comment by bluca** [comment]
also are you planning to support this in mkosi? /cc @DaanDeMeyer

**Review by bluca — APPROVED** [review]
LGTM, just two very minor things and a couple questions for future uses

**Comment by DaanDeMeyer** [comment]
> also are you planning to support this in mkosi?

I'm not entirely sure what support for this in mkosi would look like. You could just pass these as extra arguments to mkosi boot and they get forwarded to nspawn. Maybe we could add an option to configure custom nspawn args and qemu args in the config file. Those would then always get appended to the boot/qemu verb command lines respectively. Is that sufficient or do you expect something else as well?

**Comment by bluca** [comment]
> > also are you planning to support this in mkosi?
> 
> I'm not entirely sure what support for this in mkosi would look like. You could just pass these as extra arguments to mkosi boot and they get forwarded to nspawn. Maybe we could add an option to configure custom nspawn args and qemu args in the config file. Those would then always get appended to the boot/qemu verb command lines respectively. Is that sufficient or do you expect something else as well?

I didn't have any specific ideas, just a generic thought. I think there's support for setting a password at build time, maybe that could become exclusively runtime if the built/booted image is new enough?

**Comment by poettering** [comment]
> The integration with nspawn is great - how about hosts? Is there any way/plan to allow preseeded firstboot? Maybe via the kernel command line?

What creds give us is a secure way to pass the data along, i.e. we go into great lengths to protect the data as much as we can, i.e. not swap it out and set the most restrictive access perms we can. This hence i great for passing passwords and other key material. The kernel command line is not like that, anyone can read /proc/cmdine, it's world-readable. That means it sucks for passing secret stuff like root passwords. I am not aware of any sensible way we could pass stuff like that along. we'd need some parameter table that is not world-readable. But even the new-style `/proc/bootconfig` stuff is world-readable unfortunately...

**Comment by poettering** [comment]
> also are you planning to support this in mkosi? /cc @DaanDeMeyer

mkosi already knows about nspawn settings files. We probably should expose the credentials concept there too. But of course that would not work for secret creds. For those we might want to teach nspawn to read them from some separate creds bundle on disk somehow.

**Comment by poettering** [comment]
force pushed a new version with all changes suggested by @bluca

**Pull request merged**